rust/qr_reader_pc/src/lib.rs

```diff
@@ -1,104 +1,229 @@
-use rscam::{Camera, Config};
-use std::io::Write;
-use image::{GenericImageView, Pixel, Luma, ImageBuffer, GrayImage};
-use quircs;
-use hex;
-use qr_reader_phone::process_payload::{process_decoded_payload, Ready, InProgress};
+#![deny(missing_docs)]
+
+//! # QR reader crate for PC
+//!
+//! `qr_reader_pc` is a utility to capture (via webcam) QR codes from Signer
+//! and extracting data from it.
+
+use minifb::{Window, WindowOptions};
+use nokhwa::{query_devices, CaptureAPIBackend};
+use nokhwa::{Camera, CameraFormat, FrameFormat};
 use anyhow::anyhow;
+use image::{GrayImage, ImageBuffer, Pixel};
+use qr_reader_phone::process_payload::{process_decoded_payload, InProgress, Ready};
 
-const WIDTH: u32 = 640;
-const HEIGHT: u32 = 480;
+// Default camera settings
+const DEFAULT_WIDTH: u32 = 640;
+const DEFAULT_HEIGHT: u32 = 480;
+const DEFAULT_FRAME_FORMAT: FrameFormat = FrameFormat::YUYV;
+const DEFAULT_FPS: u32 = 30;
+const DEFAULT_BACKEND: CaptureAPIBackend = CaptureAPIBackend::Video4Linux;
 
-pub fn run_with_camera() -> anyhow::Result<String> {
+/// Structure for storing camera settings.
+#[derive(Debug)]
+pub struct CameraSettings {
+    index: Option<usize>,
+    frame_format: FrameFormat,
+    fps: u32,
+}
 
-    let mut camera = match Camera::new("/dev/video0") {
+/// Main cycle of video capture.
+/// Returns a string with decoded QR message in HEX format or error.
+///
+/// # Arguments
+///
+/// * `camera_settings` - CameraSettings struct that holds the camera parameters
```

Comment:
Clean the camera buffer somewhere in the beginning and in the end; camera immediately parses last frame from previous scan resulting in instant "success".

Reply:
Nokhwa crate does not have an API for working with a camera buffer (and the frame read methods are blocking), but to avoid such a problem, I can add skipping few frames at the start of the program. In an alternative PR #873 there are more opportunities to solve this problem.

Reply:
1b48054030ad98f363cbba779de5911b0295cd6f dropping frames at the beginning